File: src/geegoo_agent/runtime/executor.py
"""Tool executor (L4)."""

from __future__ import annotations

from geegoo_agent.infra.events import InProcessEventBus
from geegoo_agent.tools.registry import ToolRegistry
from geegoo_agent.tools.types import ToolCallRequest, ToolContext, ToolResult


class Executor:
    def __init__(self, registry: ToolRegistry, event_bus: InProcessEventBus | None = None) -> None:
        self._registry = registry
        self._event_bus = event_bus

    def execute(self, call: ToolCallRequest, ctx: ToolContext) -> ToolResult:
        if ctx.event_bus is None and self._event_bus is not None:
            ctx.event_bus = self._event_bus
        return self._registry.execute(call, ctx)

File: internal/runtime/executor.go
package runtime

import (
	"github.com/ghsemail/GeeGooAgent/internal/tools"
)

// Executor dispatches tool calls to the registry.
type Executor struct {
	registry *tools.Registry
}

// Registry returns the underlying tool registry.
func (e *Executor) Registry() *tools.Registry {
	if e == nil {
		return nil
	}
	return e.registry
}

// NewExecutor creates a tool executor.
func NewExecutor(registry *tools.Registry) *Executor {
	return &Executor{registry: registry}
}

// Execute runs one tool call.
func (e *Executor) Execute(req tools.CallRequest, ctx tools.Context) tools.Result {
	return e.registry.Execute(req, ctx)
}
